pkg/clusteragent/admission/mutate/autoinstrumentation/image_resolver.go

```diff
@@ -0,0 +1,271 @@
+	defer cancel()
+
+	backoff := 100 * time.Millisecond
+	maxBackoff := 2 * time.Second
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("timeout waiting for initial remote config after %v", timeout)
+		case <-time.After(backoff):
+			currentConfigs := r.rcClient.GetConfigs(state.ProductGradualRollout)
+			if len(currentConfigs) > 0 {
+				log.Infof("Loaded initial image resolution config after backoff: %d configurations", len(currentConfigs))
+				r.updateCache(currentConfigs)
+				return nil
+			}
+
+			log.Debugf("Still waiting for initial remote config, retrying in %v", backoff)
+			backoff = min(backoff*2, maxBackoff)
+		}
+	}
+}
+
+// Resolve resolves a registry, repository, and tag to a digest-based reference.
+// Input: "gcr.io/datadoghq", "dd-lib-python-init", "v3"
+// Output: "gcr.io/datadoghq/dd-lib-python-init@sha256:abc123...", true
+// If resolution fails or is not available, it returns nil.
+// Output: nil, false
+func (r *remoteConfigImageResolver) Resolve(registry string, repository string, tag string) (*ResolvedImage, bool) {
+	if !isDatadoghqRegistry(registry) {
+		log.Debugf("Not a Datadoghq registry, not resolving")
+		return nil, false
+	}
+
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if len(r.imageMappings) == 0 {
+		log.Debugf("Cache empty, no resolution available")
+		return nil, false
+	}
+
+	repoCache, exists := r.imageMappings[repository]
+	if !exists {
+		log.Debugf("No mapping found for repository %s", repository)
+		return nil, false
+	}
+
+	resolved, exists := repoCache[tag]
+	if !exists {
+		log.Debugf("No mapping found for %s:%s", repository, tag)
+		return nil, false
+	}
+
+	log.Debugf("Resolved %s:%s -> %s", repository, tag, resolved)
+	return &resolved, true
+}
+
+func isDatadoghqRegistry(registry string) bool {
+	_, exists := datadoghqRegistries[registry]
+	return exists
+}
+
+// updateCache processes configuration data and updates the image mappings cache.
+// This is the core logic shared by both initialization and remote config updates.
+func (r *remoteConfigImageResolver) updateCache(configs map[string]state.RawConfig) {
+	validConfigs, errors := parseAndValidateConfigs(configs)
+
+	for configKey, err := range errors {
+		log.Errorf("Failed to process config %s during initialization: %v", configKey, err)
+	}
+
+	r.updateCacheFromParsedConfigs(validConfigs)
+}
+
+func parseAndValidateConfigs(configs map[string]state.RawConfig) (map[string]RepositoryConfig, map[string]error) {
+	validConfigs := make(map[string]RepositoryConfig)
+	errors := make(map[string]error)
+	for configKey, rawConfig := range configs {
+		var repo RepositoryConfig
+		if err := json.Unmarshal(rawConfig.Config, &repo); err != nil {
+			errors[configKey] = fmt.Errorf("failed to unmarshal: %w", err)
+			continue
+		}
+
+		if repo.RepositoryName == "" || repo.RepositoryURL == "" {
+			errors[configKey] = fmt.Errorf("missing repository_name or repository_url in config %s", configKey)
```

Comment:
Given these are grouped by config_key and then also logged by config key by the caller, it's going to be pretty verbose, you can also consider using `multierror` here. 

Used in other parts of the agent:
https://github.com/DataDog/datadog-agent/blob/a79bec82cae278364ad29ee07cbbeeaba3f07db7/pkg/clusteragent/autoscaling/workload/config_retriever_values.go#L18